core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java

```diff
@@ -0,0 +1,189 @@
+        try (InputStream inputStream = blobContainer.openInput(blobName)) {
+            byte[] bytes = ByteStreams.toByteArray(inputStream);
+            final String resourceDesc = "ChecksumBlobStoreFormat.readBlob(blob=\"" + blobName + "\")";
+            try (ByteArrayIndexInput indexInput = new ByteArrayIndexInput(resourceDesc, bytes)) {
+                CodecUtil.checksumEntireFile(indexInput);
+                CodecUtil.checkHeader(indexInput, codec, VERSION, VERSION);
+                long filePointer = indexInput.getFilePointer();
+                long contentSize = indexInput.length() - CodecUtil.footerLength() - filePointer;
+                BytesReference bytesReference = new BytesArray(bytes, (int)filePointer, (int)contentSize);
+                return read(bytesReference);
+            } catch(CorruptIndexException | IndexFormatTooOldException | IndexFormatTooNewException ex) {
+                // we trick this into a dedicated exception with the original stacktrace
+                throw new CorruptStateException(ex);
+            }
+        }
+    }
+
+    /**
+     * Writes properly formatted blob in atomic manner.
+     * <p/>
+     * The blob will be compressed and checksum will be written if required.
+     *
+     * @param obj           object to be serialized
+     * @param blobContainer blob container
+     * @param name          blob name
+     * @throws IOException
+     */
+    public void writeAtomic(T obj, BlobContainer blobContainer, String name) throws IOException {
+        String blobName = blobName(name);
+        String tempBlobName = tempBlobName(name);
+        writeBlob(obj, blobContainer, tempBlobName);
+        blobContainer.move(tempBlobName, blobName);
+    }
+
+    /**
+     * Writes properly formatted blob in atomic manner.
+     * <p/>
+     * The blob will be compressed and checksum will be written if required.
+     *
+     * @param obj           object to be serialized
+     * @param blobContainer blob container
+     * @param name          blob name
+     * @throws IOException
+     */
+    public void write(T obj, BlobContainer blobContainer, String name) throws IOException {
+        String blobName = blobName(name);
+        writeBlob(obj, blobContainer, blobName);
+    }
+
+    public void writeBlob(T obj, BlobContainer blobContainer, String blobName) throws IOException {
+        BytesReference bytes = write(obj);
+        try (OutputStream outputStream = blobContainer.createOutput(blobName)) {
+            final String resourceDesc = "ChecksumBlobStoreFormat.writeBlob(blob=\"" + blobName + "\")";
+            try (OutputStreamIndexOutput indexOutput = new OutputStreamIndexOutput(resourceDesc, outputStream, BUFFER_SIZE)) {
+                CodecUtil.writeHeader(indexOutput, codec, VERSION);
+                try (OutputStream indexOutputOutputStream = new org.elasticsearch.common.lucene.store.OutputStreamIndexOutput(indexOutput) {
```

Comment:
These are two different class with the same name. The first `OutputStreamIndexOutput ` 3 lines above is `org.apache.lucene.store.OutputStreamIndexOutput`. It is `IndexOutput` facade that writes into `OutputStream`. 

The second `OutputStreamIndexOutput` with fully-qualified name is `OutputStream` facade that writes into `IndexOutput`. 

All this makes me thing that `org.elasticsearch.common.lucene.store.OutputStreamIndexOutput` should be renamed to `IndexOutputOutputStream`. I will do that.